homeassistant/components/mpris/config_flow.py

```diff
@@ -0,0 +1,405 @@
+        )
+
+        # Do not proceed if the device is already configured by hand.
+        await self.async_set_unique_id(self._regen_unique_id())
+        self._abort_if_unique_id_configured()
+
+        return await self.async_step_pairing()
+
+    async def async_step_zeroconf(
+        self, discovery_info: zeroconf.ZeroconfServiceInfo
+    ) -> FlowResult:
+        """Handle zeroconf discovery."""
+        if (
+            not discovery_info.host
+            or not discovery_info.port
+            or not discovery_info.hostname
+            or not discovery_info.name
+            or not discovery_info.properties
+            or CONF_CAKES_PORT not in discovery_info.properties
+        ):
+            _LOGGER.debug("Ignoring invalid zeroconf announcement: %s", discovery_info)
+            return self.async_abort(reason=REASON_INVALID_ZEROCONF)
+
+        split_discovery = discovery_info.name.split(".")
+        self._set_data(
+            split_discovery[0],
+            discovery_info.host,
+            int(discovery_info.properties[CONF_CAKES_PORT]),
+            int(discovery_info.port or DEF_MPRIS_PORT),
+        )
+
+        # Do not proceed if the device is already configured by hand
+        # or by zeroconf.
+        # User may have already added the agent by hostname manually,
+        # so we don't want to show this to the user again since the
+        # discovery is yielding a machine already added.
+        # There is a corner case we are not dealing with here, and that
+        # corner case is when the zeroconf hostname / port data has changed,
+        # indicating that the machine has changed networking, or the user
+        # of the agent machine has logged after someone else logged in
+        # (under such circumstance, the agent may decide to allocate
+        # different CAKES and MPRIS ports).
+        # Ideally, in this situation, we would detect that the agent is
+        # the same (by using some signed message in the zeroconf properties,
+        # or enforcing that the certificate fingerprint in the zeroconf
+        # properties is the same as the one we know about, which would
+        # require us to use the fingerprint as the unique ID, which is OK),
+        # then proceed to update the hostname, CAKES and MPRIS ports of any
+        # existing config entry, and reload the config entry in question.
+        # This is a corner case to be dealt with in the future rather than
+        # now, since it would require a mechanism for the integration to
+        # trust that the networking information change announced by the
+        # agent is in fact legitimate and corresponds to the existing
+        # config entry's certificate.  In the future, the zeroconf
+        # announcement will send a signed message that this integration
+        # can verify using the relevant certificate.
+        await self.async_set_unique_id(self._regen_unique_id())
+        self._abort_if_unique_id_configured()
```

Comment:
> This also means a config entry update listener is needed in `__init__.py`

what

Sorry, this is confusing.  Not sure what a config entry update listener is.

What would be ideal is that zeroconf doesn't capture just host change — because the ports could also change — but rather captures a whole new set of host/port/port data based on a hash unique to the agent (and that would be the hash of the server certificate the agent has).

Right now on host updates we simply generate a completely different unique ID, and that just appears as a new device.

I think until the agent side provides the hash of the cert as a zeroconf data key/value pair, what we have today is acceptable.  We do not expect IP address change to be happening regularly, and for those users who do have this circumstance, they can always manually add an entry by using a dynamic host name.  When the server side (agent) is ready to advertise its server certificate ID, then we'll get to the point of using that as the unique ID, and then both manual and zeroconf config entries will just work.

Reply:
Leaving this unresolved because this conversation is very interesting and I can definitely stand to learn more about this.

Reply:
To expedite the process, I have changed the code to remove the comment and to simply use host/port unique ID generation.  I will leave a short comment explaining that, in the future, we will use certificate fingerprint revalidation instead.

My offer to hold on and implement fingerprint-based revalidation stands!  Meanwhile, I'm closing this comment.